source/extensions/dynamic_modules/abi.h

```diff
@@ -2726,6 +2906,249 @@ bool envoy_dynamic_module_callback_network_filter_is_ssl(
+ */
+bool envoy_dynamic_module_callback_listener_filter_get_buffer_slice(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_envoy_ptr* data_ptr_out, size_t* data_length_out);
+
+/**
+ * envoy_dynamic_module_callback_listener_filter_drain_buffer is called by the module to drain
+ * bytes from the beginning of the buffer.
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleListenerFilter object.
+ * @param length is the number of bytes to drain from the beginning of the buffer.
+ * @return true if the drain was successful, false otherwise.
+ */
+bool envoy_dynamic_module_callback_listener_filter_drain_buffer(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr, size_t length);
+
+// -----------------------------------------------------------------------------
+// Socket Property Setters (Protocol Detection)
+// -----------------------------------------------------------------------------
+
+/**
+ * envoy_dynamic_module_callback_listener_filter_set_detected_transport_protocol is called by the
+ * module to set the detected transport protocol (e.g., "tls", "raw_buffer").
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleListenerFilter object.
+ * @param protocol is the pointer to the protocol string.
+ * @param length is the length of the protocol string.
+ */
+void envoy_dynamic_module_callback_listener_filter_set_detected_transport_protocol(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_const_buffer_envoy_ptr protocol, size_t length);
+
+/**
+ * envoy_dynamic_module_callback_listener_filter_set_requested_server_name is called by the module
+ * to set the requested server name (SNI).
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleListenerFilter object.
+ * @param name is the pointer to the server name string.
+ * @param length is the length of the server name string.
+ */
+void envoy_dynamic_module_callback_listener_filter_set_requested_server_name(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_const_buffer_envoy_ptr name, size_t length);
+
+/**
+ * envoy_dynamic_module_callback_listener_filter_set_requested_application_protocols is called by
+ * the module to set the requested application protocols (ALPN).
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleListenerFilter object.
+ * @param protocols_ptr is an array of protocol name pointers.
+ * @param protocols_length is an array of protocol name lengths.
+ * @param protocols_count is the number of protocols.
+ */
+void envoy_dynamic_module_callback_listener_filter_set_requested_application_protocols(
+    envoy_dynamic_module_type_listener_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_const_buffer_envoy_ptr* protocols_ptr, size_t* protocols_length,
+    size_t protocols_count);
```

Comment:
You should use a pointer/array of `envoy_dynamic_module_type_module_buffer` and an array length here.